<launch>

    <node name="sampleimugpspublisher2SVY21" pkg="epochmini" type="sampleimugpspublisher2SVY21.py" />

    <node name="neato_laser_publisher" pkg="xv_11_laser_driver" type="neato_laser_publisher">
    <param name="port" value="/dev/ttyACM1" type="str" />
    <param name="firmware_version" value="2" type="int" />
  </node>

    <node pkg="robot_localization" type="ukf_localization_node" name="ukf_odom" clear_params="true">
      <remap from="odometry/filtered" to="odom/ukf"/>

      <param name="frequency" value="30"/>
      <param name="sensor_timeout" value="0.1"/>
      <param name="two_d_mode" value="true"/>

      <param name="map_frame" value="world"/>
      <param name="odom_frame" value="odom"/>
      <param name="base_link_frame" value="base_link"/>
      <param name="world_frame" value="odom"/>

      <param name="transform_time_offset" value="0.0"/>

      <param name="imu0" value="imuBoat"/>
      <param name="odom0" value="odomGPS"/>

      <rosparam param="imu0_config">[false, false, false,
                                     false,  false,  true,
                                     false, false, false,
                                     false,  false,  true,
                                     true,  true,  false]</rosparam>

      <rosparam param="odom0_config">[false, false,  false,
                                      false, false, false,
                                      true, true, false,
                                      false, false, true,
                                      false, false, false]</rosparam>

      <param name="imu0_differential" value="false"/>
      <param name="odom0_differential" value="true"/>
      
      <param name="imu0_relative" value="true"/>
      <param name="odom0_relative" value="false"/>

      <param name="imu0_remove_gravitational_acceleration" value="true"/>

      <param name="print_diagnostics" value="true"/>

      <!-- ======== ADVANCED PARAMETERS ======== -->

      <param name="imu0_queue_size" value="20"/>
      <param name="twist0_queue_size" value="5"/>

      <param name="imu0_pose_rejection_threshold" value="0.3"/>
      <param name="imu0_twist_rejection_threshold" value="0.1"/>
      <param name="imu0_linear_acceleration_rejection_threshold" value="0.1"/>
      <param name="odom0_pose_rejection_threshold" value="5"/>
      <param name="odom0_twist_rejection_threshold" value="1"/>

      <param name="debug"           value="false"/>
      <param name="debug_out_file"  value="debug_ukf_localization.txt"/>

      <!-- ======== UKF-SPECIFIC PARAMETERS ======== -->

      <param name="alpha" value="0.001"/>
      <param name="kappa" value="0"/>
      <param name="beta" value="2"/>

    </node>

    <node pkg="robot_localization" type="ukf_localization_node" name="ukf_world" clear_params="true">
      <remap from="odometry/filtered" to="world/ukf"/>

      <param name="frequency" value="30"/>
      <param name="sensor_timeout" value="0.1"/>
      <param name="two_d_mode" value="true"/>

      <param name="map_frame" value="world"/>
      <param name="odom_frame" value="odom"/>
      <param name="base_link_frame" value="base_link"/>
      <param name="world_frame" value="world"/>

      <param name="transform_time_offset" value="0.0"/>

      <param name="imu0" value="imuBoat"/>
      <param name="odom0" value="odomGPS"/>
      <param name="pose0" value="poseGPS"/>

      <rosparam param="imu0_config">[false, false, false,
                                     false,  false,  true,
                                     false, false, false,
                                     false,  false,  true,
                                     true,  true,  false]</rosparam>

      <rosparam param="odom0_config">[false, false,  false,
                                      false, false, false,
                                      true, true, false,
                                      false, false, true,
                                      false, false, false]</rosparam>

      <rosparam param="pose0_config">[true,  true,  false,
                                      false, false, false,
                                      false, false, false,
                                      false, false, false,
                                      false, false, false]</rosparam>

      <param name="imu0_differential" value="false"/>
      <param name="odom0_differential" value="true"/>
      <param name="pose0_differential" value="true"/>
      
      <param name="imu0_relative" value="true"/>
      <param name="odom0_relative" value="false"/>
      <param name="pose0_relative" value="false"/>

      <param name="imu0_remove_gravitational_acceleration" value="true"/>

      <param name="print_diagnostics" value="true"/>

      <!-- ======== ADVANCED PARAMETERS ======== -->

      <param name="imu0_queue_size" value="20"/>
      <param name="odom0_queue_size" value="5"/>
      <param name="pose0_queue_size" value="5"/>

      <param name="imu0_pose_rejection_threshold" value="0.3"/>
      <param name="imu0_twist_rejection_threshold" value="0.1"/>
      <param name="imu0_linear_acceleration_rejection_threshold" value="0.1"/>
      <param name="odom0_pose_rejection_threshold" value="5"/>
      <param name="odom0_twist_rejection_threshold" value="1"/>
      <param name="imu0_pose_rejection_threshold" value="0.3"/>

      <param name="debug"           value="false"/>
      <param name="debug_out_file"  value="debug_ukf_localization.txt"/>

      <!-- ======== UKF-SPECIFIC PARAMETERS ======== -->

      <param name="alpha" value="0.001"/>
      <param name="kappa" value="0"/>
      <param name="beta" value="2"/>

    </node>

    <node name="tf_boattoneato_publisher" pkg="tf" type="static_transform_publisher" args="0.3 0 0 0 0 0 1 boat_GPS neato_laser 10" />

    <node name="gmapping" pkg="gmapping" type="slam_gmapping" output="screen">
      <<!-- remap from="scan" to="base_scan"/> -->
      <param name="map_update_interval" value="5.0"/>
      <param name="maxUrange" value="5.0"/>
      <param name="sigma" value="0.05"/>
      <param name="kernelSize" value="1"/>
      <param name="lstep" value="0.05"/>
      <param name="astep" value="0.05"/>
      <param name="iterations" value="5"/>
      <param name="lsigma" value="0.075"/>
      <param name="ogain" value="3.0"/>
      <param name="lskip" value="0"/>
      <param name="srr" value="0.1"/>
      <param name="srt" value="0.2"/>
      <param name="str" value="0.1"/>
      <param name="stt" value="0.2"/>
      <param name="linearUpdate" value="1.0"/>
      <param name="angularUpdate" value="0.5"/>
      <param name="temporalUpdate" value="3.0"/>
      <param name="resampleThreshold" value="0.5"/>
      <param name="particles" value="30"/>
      <param name="xmin" value="-50.0"/>
      <param name="ymin" value="-50.0"/>
      <param name="xmax" value="50.0"/>
      <param name="ymax" value="50.0"/>
      <param name="delta" value="0.05"/>
      <param name="llsamplerange" value="0.01"/>
      <param name="llsamplestep" value="0.01"/>
      <param name="lasamplerange" value="0.005"/>
      <param name="lasamplestep" value="0.005"/>
</node>

</launch>
